h5io_browser/base.py

```diff
@@ -229,6 +230,44 @@ def _merge_nested_dict(main_dict: dict, add_dict: dict) -> dict:
             main_dict[k] = v
     return main_dict
 
+from collections import defaultdict
+_FILE_HANDLE_CACHE_COUNTER = defaultdict(int)
+_FILE_HANDLE_CACHE = {}
+import threading
+_FILE_HANDLE_CACHE_LOCK = threading.RLock()
+
+import contextlib
+@contextlib.contextmanager
+def WithOpenHDF(filename: str, mode: str = "r", swmr: bool = False):
+    filename = os.path.realpath(os.path.abspath(filename))
+    with _FILE_HANDLE_CACHE_LOCK:
+        file_is_open = filename in _FILE_HANDLE_CACHE
+    # File already cached, someone above us opened it and will clean it up, we just need to pass it down
+    if file_is_open:
+        # print("EVERYBODY HIT THE CACHE")
+        with _FILE_HANDLE_CACHE_LOCK:
+            handle = _FILE_HANDLE_CACHE[filename]
+        assert mode == handle.mode
+        if swmr != handle.swmr_mode:
+            print("WARNING: Found a open handle, but swmr mode different!")
+        yield handle
+    else:
+        with _open_hdf(filename, mode, swmr) as handle:
+            with _FILE_HANDLE_CACHE_LOCK:
+                _FILE_HANDLE_CACHE[filename] = handle
+            try:
+                # Monkey patch close so that downstream code can use the handle as a context manager without closing the
+                # file we want to cache
+                handle.close = lambda: None
```

Comment:
This will not help.

- If the competing close is from an outer context, there is no need, because the outer context will close the file after the inner context has ended. 
- Closing the file manually within the context is against intended use.
- Protecting against close from concurrent threads works better via reference counting. Manipulation of handle is not thread-safe, so race-conditions are garantueed.

Reply:
> * If the competing close is from an outer context, there is no need, because the outer context will close the file after the inner context has ended. 
> * Closing the file manually within the context is against intended use.

The issue is code that is unaware that the context is used (somewhere up the call stack).  Say I have two functions `foo` and `bar`.  The use the `_open_hdf` function to obtain a handle and use it as a context manager, because they want to make sure the file is closed.
```python
def foo(...):
  with _open_hdf() as handle:
    ...
def bar(...):
  with _open_hdf() as handle:
    ...
```
(I'm leaving out the details but assume they open the same file).
If I call them within the new context 
```python
with WithOpenHDF():
  foo()
  bar()
```
both calls receive the same handle from `_FILE_HANDLE_CACHE`, but the `with ... as handle:` in `foo()` will close the handle after the block. Our context manager doesn't get run though, so we cannot update the cache (and if we did it would subvert the whole point of the cache), which means `bar()` will get a cached but closed file handle.

> Protecting against close from concurrent threads works better via reference counting. Manipulation of handle is not thread-safe, so race-conditions are garantueed.

Agree, but the only way without monkey patching `close` I see is making a wrapping proxy object that behaves just like a hdf handle, except it has a reference counting `close`.  Would be doable, but annoying.  Maybe giving up on thread safety is the saner option then.